scrapy_autoextract/providers.py

```diff
@@ -0,0 +1,98 @@
+from typing import ClassVar, Optional, Type
+
+from autoextract.aio import request_raw
+from autoextract.request import Request as AutoExtractRequest
+from autoextract_poet.page_inputs import (
+    AutoExtractArticleData,
+    AutoExtractProductData,
+)
+from scrapy import Request
+from scrapy.settings import Settings
+from scrapy.statscollectors import StatsCollector
+from scrapy_poet.page_input_providers import (
+    PageObjectInputProvider,
+    register,
+)
+
+
+class _Provider(PageObjectInputProvider):
+    """An interface that describes a generic AutoExtract Provider.
+
+    It should not be used publicly as it serves the purpose of being a base
+    class for more specific providers such as Article and Product providers.
+    """
+
+    provided_class: ClassVar[Optional[Type]]
```

Comment:
Why `Optional`? Shouldn't be always there?

Reply:
The type must be a subclass of `_AutoExtractData` ensuring that item_key is declared. It could be nice to have a comment here stating that.

Reply:
> Why Optional? Shouldn't be always there?

It shouldn't be optional. I've just fixed it. Thanks.

Reply:
> The type must be a subclass of _AutoExtractData ensuring that item_key is declared.

We could do something like `provided_class: ClassVar[Type[_AutoExtractData]]` if this class was not private. 

> It could be nice to have a comment here stating that.

I guess I'm going with this option. Let me know if you, @kmike or @ejulio have another option.

Reply:
What do you think? Should we make `_AutoExtractData` public and use it here for describing `provided_class`'s type?

Reply:
+1 to add a comment

Reply:
+1 to add a comment.